internal/backend/remote-state/azure/auth/clients.go

```diff
@@ -0,0 +1,227 @@
+			tfdiags.Error,
+			"Resource Group is empty",
+			"In order to obtain a Storage Account Access Key, a resource group is necessary",
+		))
+	}
+	if names.StorageAccount == "" {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Storage Account is empty",
+			"In order to obtain a Storage Account Access Key, a storage account name is necessary",
+		))
+	}
+	if names.SubscriptionID == "" {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Subscription ID is empty",
+			"In order to obtain a Storage Account Access Key, a subscription id is necessary",
+		))
+	}
+	return diags.Err()
+}
+
+// NewContainerClientWithSharedKeyCredentialAndKey gets a container client and shared key
+// that it's authenticated with. This function should only be used for testing and internally within this package.
+func NewContainerClientWithSharedKeyCredentialAndKey(ctx context.Context, names StorageAddresses, authCred azcore.TokenCredential) (*container.Client, string, error) {
+	client := httpclient.New(ctx)
+	// Lookup the key with an account client
+	accountsClient, err := NewStorageAccountsClient(client, authCred, names.CloudConfig, names.SubscriptionID)
+	if err != nil {
+		return nil, "", err
+	}
+	keys, err := accountsClient.ListKeys(ctx, names.ResourceGroup, names.StorageAccount, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("error listing access keys on the storage account: %w", err)
+	}
+	if len(keys.Keys) == 0 || keys.Keys[0] == nil || keys.Keys[0].Value == nil {
+		return nil, "", errors.New("malformed structure returned from the ListKeys function")
+	}
+
+	storageAccessKey := *keys.Keys[0].Value
+
+	return newContainerClientFromStorageAccessKey(client, names, storageAccessKey)
+}
+
+func CloudConfigFromAddresses(ctx context.Context, environment, metadataHost string) (cloud.Configuration, error) {
+	if metadataHost != "" {
+		return CloudConfigFromMetadataHost(ctx, metadataHost)
+	}
+	switch environment {
+	case "", "public", "global", "canary":
+		return cloud.AzurePublic, nil
+	case "usgovernment", "usgovernmentl4", "dod", "usgovernmentl5":
+		return cloud.AzureGovernment, nil
+	case "china":
+		return cloud.AzureChina, nil
+	}
+	return cloud.Configuration{}, fmt.Errorf("unknown environment identifier: %s", environment)
+}
+
+type Authentication struct {
+	LoginEndpoint string `json:"loginEndpoint"`
+}
+
+type Environment struct {
+	Authentication  Authentication `json:"authentication"`
+	ResourceManager string         `json:"resourceManager"`
+}
+
+func CloudConfigFromMetadataHost(ctx context.Context, metadataHost string) (cloud.Configuration, error) {
+	// Obtaining cloud config from the metadata host
+	client := httpclient.New(ctx)
+	uri := fmt.Sprintf("https://%s/metadata/endpoints?api-version=2023-11-01", metadataHost)
+	req, err := http.NewRequestWithContext(ctx, "GET", uri, nil)
+	if err != nil {
+		return cloud.Configuration{}, err
+	}
+	resp, err := client.Do(req)
```

Comment:
Why not client.Get() instead of NewRequestWithContext?  The client above already has the context afaik.

Reply:
Also, don't we need to defer resp.Body.Close()?